### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Equivalence

### PR DESCRIPTION
**Pull request by hargoniX**
This is missing slice_lhs which is only used in two proofs though.

Update: `slice_lhs` (and other slice tactics) have now been ported

---

- [x] depends on: #2109

### CODE REVIEW

**Comment by joelriou** [comment]
In my understanding, `by obviously` should be replaced by `by aesop_cat`.

**Comment by hargoniX** [comment]
Okay I pushed the port as far as I could now. There is still a couple of issues remaining of course:
- the slice_lhs tactic is missing so some lemmas are not proven
- the symm calls in the file dont work for some reason despite operating on equality if I am seeing that correctly
- some simp calls don't work either, this might be because some other lemmata in the file are still not proven or other reasons I'm not sure.
- towards the end of the file there is a bunch of slightly confusing stuff with instances that I'd like to take a look at when the above things are fixed.

**Comment by mattrobball** [comment]
Current status: 
- now depends on `slice` PR #2109 
- `aesop_cat` looking good so far
- `slice_lhs` now `sliceLHS` and working 
- added a `Trans Iso Iso Iso` to use `calc` for a lemma 
- `@[refl]` complaining and one proof that seems actual thought

**Comment by mattrobball** [comment]
All green now.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2109
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2109~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by hargoniX** [comment]
@semorrison So we can merge this now I guess right?

File: Mathlib/CategoryTheory/Equivalence.lean [hunk 1]
```diff
@@ -0,0 +1,816 @@
...
+#align category_theory.equivalence.inv_fun_map CategoryTheory.Equivalence.inv_fun_map
+
+section
+
+-- In this section we convert an arbitrary equivalence to a half-adjoint equivalence.
+variable {F : C ⥤ D} {G : D ⥤ C} (η : 𝟭 C ≅ F ⋙ G) (ε : G ⋙ F ≅ 𝟭 D)
+
+/-- If `η : 𝟭 C ≅ F ⋙ G` is part of a (not necessarily half-adjoint) equivalence, we can upgrade it
+to a refined natural isomorphism `adjointify_η η : 𝟭 C ≅ F ⋙ G` which exhibits the properties
+required for a half-adjoint equivalence. See `equivalence.mk`. -/
```
Comment (adamtopaz) [inline Mathlib/CategoryTheory/Equivalence.lean:258]:
```suggestion
required for a half-adjoint equivalence. See `Equivalence.mk`. -/
```

I'm not sure what the new naming conventions say about lemma names such as `equivalence_mk'_foo`. Should these be changed to `Equivalence_mk'_foo`?

Comment (mattrobball) [reply Mathlib/CategoryTheory/Equivalence.lean:258]: You are right. I didn’t run the new comment naming convention fixer on this. So good might not be the only incident

File: Mathlib/CategoryTheory/Iso.lean [hunk 2]
```diff
@@ -140,6 +140,8 @@ def refl (X : C) : X ≅ X where
 
 instance : Inhabited (X ≅ X) := ⟨Iso.refl X⟩
 
+theorem nonempty_iso_refl (X : C) : Nonempty (X ≅ X) := ⟨default⟩
```
Comment (rwbarton) [inline Mathlib/CategoryTheory/Iso.lean:143]: Is this needed for something?

File: Mathlib/CategoryTheory/Equivalence.lean (hunk 1 above)
Comment (hargoniX) [reply Mathlib/CategoryTheory/Equivalence.lean]: Okay I think I got all relevant ones now. There is a few theorems that have names along the lines of `of_equivalence` but take an `IsEquivalence` argument so I wasn't sure whether I want to mention the name of a struct there?

File: Mathlib/CategoryTheory/Iso.lean (hunk 2 above)
Comment (hargoniX) [reply Mathlib/CategoryTheory/Iso.lean:143]: No, things do seem to build without this lemma, shall I go ahead and remove it?

Comment (mattrobball) [reply Mathlib/CategoryTheory/Iso.lean:143]: Originally this https://github.com/leanprover-community/mathlib4/blob/c49e83649a214bc334ae6aa5bef9dd7518464b0c/Mathlib/CategoryTheory/Equivalence.lean#L798 was `by simp`. Using `simp?`, this was missing some inhabited to nonempty result. I thought I put this in there to provide `default` for a fix by hand. But if it builds fine, perhaps I am mistaken. Regardless, if it builds, it is fine to me.

Comment (adamtopaz) [reply Mathlib/CategoryTheory/Iso.lean:143]:
```
⟨e.symm Y, by simpa using ⟨default⟩⟩
```
seems closer to the mathlib3 version, and it should work just with the inhabited instance.

File: Mathlib/CategoryTheory/Equivalence.lean (hunk 1 above)
Comment (adamtopaz) [reply Mathlib/CategoryTheory/Equivalence.lean]: I think that while the port is going on, we should keep the names as close as possible to their mathlib3 counterparts. So the names look good to me after your last change.

File: Mathlib/CategoryTheory/Iso.lean (hunk 2 above)
Comment (mattrobball) [reply Mathlib/CategoryTheory/Iso.lean:143]: Pushed this.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**